clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```
@@ -792,12 +800,17 @@ private void timeoutCallsInFlight(TimeoutProcessor processor, Map<String, List<C
                 // only one we need to check the timeout for.
                 Call call = contexts.get(0);
                 if (processor.callHasExpired(call)) {
-                    log.debug("{}: Closing connection to {} to time out {}", clientId, nodeId, call);
-                    client.disconnect(nodeId);
-                    numTimedOut++;
-                    // We don't remove anything from the callsInFlight data structure.  Because the connection
-                    // has been closed, the calls should be returned by the next client#poll(),
-                    // and handled at that point.
+                    if (call.aborted) {
+                        log.warn("{}: aborted call {} is still in callsInFlight.", clientId, call);
```

In the PR description, you said:

```text
It also adds the Call#aborted flag, which ensures that we only abort a connection once-- even if there is a similar bug in the future which causes old Call objects to linger.
```

It wasn't totally clear to me if you were suggesting that `abort` was useful in case we end up with a similar bug or if it was necessary for correctness under normal behaviour. It looks like the latter, is that correct?

---

It has two roles right now: ensuring that we send a TimeoutException rather than a DisconnectException when an RPC has timed out, and preventing similar bugs in the future.

---

I see. Wouldn't it be better to fail fast if we introduce a similar bug in the future instead of logging a warning? That way, we'll hopefully trigger a test failure and fix it.

---

Hmm.  I'm having trouble finding a clear way to fail fast here.  The call has already been aborted-- aborting it again should not help.  We could try to initiate an adminclient close, but this is a very heavy hammer to use and might make the situation worse for the operator.

---

Makes sense, thanks for the explanation.